<title>SPAdes Genome Assembler changelog</title>
<body>

<h2>SPAdes Genome Assembler changelog</h2>

<h3>SPAdes 3.1.1, 27 May 2014</h3>

<p>FIX: Several improvements in IonHammer.</p>

<p>FIX: Fixed a few minor bugs in repeat resolution and scaffolding.</p>

<h3>SPAdes 3.1.0, 27 May 2014</h3>

<p>NEW: Mate-pair only assembly with high-quality libraties.</p>

<p>NEW: Support for BAM files.</p>

<p>CHANGE: Improved IonTorrent pipeline.</p>

<p>CHANGE: Better quality and higher performance when using mate-pairs.</p>

<p>FIX: Fixed dipSPAdes bugs and user interface.</p>

<h3>SPAdes 3.0.0, 29 December 2013</h3>

<p>NEW: Module for assemblying diplod highly polymorphic genomes.</p>

<p>NEW: Support for PacBio reads.</p>

<p>NEW: Support for IonTorrent reads.</p>

<p>NEW: Support for Sanger reads and additional contigs.</p>

<p>NEW: Possibility to restart SPAdes starting from the specified check-point with the <code>--restart-from</code> option.</p>

<p>NEW: Output contigs/scaffolds in FASTA and FASTG.</p>

<p>CHANGE: Improved algorithm for mate-pair repeat resolution and scaffolding.</p>

<p>CHANGE: Improved N50 and misassembly rate for single-cell data sets with low genome fraction.</p>

<p>FIX: User-friendly handling for errors in mismatch corrector.</p>

<p>REMOVE: Rectangle graph repeat resolution module.</p>

<h3>SPAdes 2.5.1, 10 September 2013</h3>

<p>NEW: Python 3.2 and 3.3 compatibility.</p>

<p>NEW: Possibility to continue SPAdes run starting from the last check-point with the <code>--continue</code> option.</p>

<p>CHANGE: Decreased memory consumption for error correction module.</p>

<p>CHANGE: Improved misassembly rate for single-cell data sets with low genome fraction.</p>

<p>FIX: User-friendly handling for the case when paired reads do not align to the assembly graph.</p>

<h3>SPAdes 2.5.0, 2 July 2013</h3>

<p>NEW: Multiple paired-end and mate-pair libraries.</p>

<p>NEW: Recipe for assembling Illumina 2x150bp and 2x250bp reads.</p>

<p>CHANGE: Improved mismatch and indel rate.</p>

<h3>SPAdes 2.4.0, 26 February 2013</h3>

<p>NEW: Mismatch correction post-processing module.</p>

<p>NEW: Rectangle graph repeat resolution module as an option.</p>

<p>NEW: Build for Mac OS.</p>

<p>CHANGE: Improved assembly quality of standard (isolate) data sets.</p>

<p>CHANGE: Decreased memory consumption in error correction module (14 Gb instead of 24 Gb on E.coli test dataset).</p>

<p>REMOVE: SAM-file generation.</p>

<h3>SPAdes 2.3.0, 30 October 2012</h3>

<p>NEW: Generate scaffolds alongside with contigs.</p>

<p>CHANGE: Use N instead of A, C, G, T for the variations in repeats.</p>

<p>CHANGE: Memory requirements for E.coli test dataset decreased from 35 Gb to 24 Gb of RAM.</p>

<p>CHANGE: output_dir is a required command line parameter instead of project_name.</p>

<p>CHANGE: Simplified output directory structure.</p>

<p>CHANGE: CMake 2.8 is required instead of 2.6.</p>

<p>REMOVE: No dependency from boost library.</p>

<h3>SPAdes 2.2.1, 20 August 2012</h3>

<p>FIX: Avoid <code>Verification of expression 'v1 == conjugate(v2)' failed</code> error.</p>


<h3>SPAdes 2.2.0, 2 August 2012</h3>

<p>NEW: No special binaries for different K values.</p>

<p>NEW: Great improvements in error correction tool BayesHammer.</p>

<p>CHANGE: Memory requirements for E.coli test dataset decreased from 85 Gb to 35 Gb of RAM.</p>

<p>CHANGE: Only 1 iteration of BayesHammer by default.</p>

<p>NEW: Improved assembly quality.</p>

<h3>SPAdes 2.1.0, 28 May 2012</h3>

<p>NEW: Support multi-threading.</p>

<p>NEW: Improved algorithms.</p>

<p>NEW: Command-line interface.

<p>CHANGE: Quality assesment separated from the core pipeline.</p>

<p>REMOVE: No support for debian and RPM packages.</p>

<h3>SPAdes 2.0.1, 26 Apr 2012</h3>

<p>FIX: Quality tool fixed.</p>

<h3>SPAdes 2.0.0, 18 Apr 2012</h3>

<p>Initial release.</p>

<p>
    <a href="http://bioinf.spbau.ru/en/spades">SPAdes site</a>
</p>

</body>
</html>

